control-plane/agents/src/bin/core/volume/registry.rs

```diff
@@ -247,6 +245,71 @@ impl Registry {
             }
         }
 
-        Ok(VolumeSnapshotState::new(snapshot, size, replica_snaps))
+        VolumeSnapshotState::new(snapshot, size, replica_snaps)
+    }
+
+    /// Return a single snapshot object corresponding to the either matching (volumeid + snapid)
+    /// OR matching snapid when input volumeid is None.
+    pub(crate) async fn snapshot(
+        &self,
+        vol_id: Option<&VolumeId>,
+        snap_id: &SnapshotId,
+    ) -> Result<grpc_mod::VolumeSnapshot, SvcError> {
+        // Since snapshot id is universally unique, we can use only snap id to filter the spec.
+        let snapshot =
+            self.specs()
+                .volume_snapshot_rsc(snap_id)
+                .ok_or(SvcError::VolSnapshotNotFound {
+                    snap_id: snap_id.to_string(),
+                    source_id: vol_id.map(|v| v.to_string()),
+                })?;
+
+        let snap_source_id = snapshot.lock().spec().source_id().clone();
+
+        // Verify that the fetched snashot spec has the same volume id as input.
+        // For debug builds assert, for production builds return error that indicates
+        // the provided input combination of volume id and snapshot id isn't valid.
+        if let Some(vid) = vol_id {
+            debug_assert!(&snap_source_id == vid);
+            return Err(SvcError::InvalidSnapshotSource {
```

Comment:
hmm shouldn't we have a check here? Seems like it's always returning this error as long as vol_id is Some.

Reply:
right, this must be:

```
if let Some(vid) = vol_id {
       if vid != &snap_source_id {
            return Err...
        }
 }
```

Reply:
has to be a separate if statement I think